packages/flutter/lib/src/services/hardware_keyboard.dart

```diff
@@ -0,0 +1,894 @@
+    }
+  }
+
+  bool _dispatchKeyEvent(KeyEvent event) {
+    // This dispatching could have used the same algorithm as [ChangeNotifier],
+    // but since 1) it shouldn't be necessary to support reentrantly
+    // dispatching, 2) there shouldn't be many handlers (most apps should use
+    // only 1, this function just uses a simpler algorithm.
+    assert(!_duringDispatch, 'Nested keyboard dispatching is not supported');
+    _duringDispatch = true;
+    bool handled = false;
+    for (final KeyEventCallback handler in _handlers) {
+      try {
+        final bool thisResult = handler(event);
+        handled = handled || thisResult;
+      } catch (exception, stack) {
+        FlutterError.reportError(FlutterErrorDetails(
+          exception: exception,
+          stack: stack,
+          library: 'services library',
+          context: ErrorDescription('while dispatching notifications for $runtimeType'),
+          informationCollector: () sync* {
+            yield DiagnosticsProperty<HardwareKeyboard>(
+              'The $runtimeType sending notification was',
+              this,
+              style: DiagnosticsTreeStyle.errorProperty,
+            );
+          },
+        ));
+      }
+    }
+    _duringDispatch = false;
+    if (_modifiedHandlers != null) {
+      _handlers = _modifiedHandlers!;
+      _modifiedHandlers = null;
+    }
+    return handled;
+  }
+
+  /// Process a new [KeyEvent] by recording the state changes and dispatching
+  /// to handlers.
+  bool handleKeyEvent(KeyEvent event) {
+    _assertEventIsRegular(event);
+    final PhysicalKeyboardKey physicalKey = event.physicalKey;
+    final LogicalKeyboardKey logicalKey = event.logicalKey;
+    if (event is KeyDownEvent) {
+      _pressedKeys[physicalKey] = logicalKey;
+      final KeyboardLockMode? lockMode = KeyboardLockMode.findLockByLogicalKey(event.logicalKey);
+      if (lockMode != null) {
+        if (_lockModes.contains(lockMode)) {
+          _lockModes.remove(lockMode);
+        } else {
+          _lockModes.add(lockMode);
+        }
+      }
+    } else if (event is KeyUpEvent) {
+      _pressedKeys.remove(physicalKey);
+    } else if (event is KeyRepeatEvent) {
+      // Empty
+    }
+
+    return _dispatchKeyEvent(event);
+  }
+
+  /// Clear all keyboard states and additional handlers.
+  ///
+  /// All handlers are removed except for the 1st one, which is added by
+  /// [ServicesBinding].
+  ///
+  /// This is used by the testing framework to make sure tests are hermetic.
+  @visibleForTesting
+  void clearState() {
+    assert(() {
+      _pressedKeys.clear();
+      _lockModes.clear();
+      _handlers.clear();
+      assert(_modifiedHandlers == null);
+      return true;
+    }());
+  }
+}
+
+/// The mode in which information of key events is delivered.
+///
+/// Different platforms deliver key events in different modes, which differ
+/// in dispatching functions, event classes, and event models. These modes are
+/// called "key event vehicles".
```

Comment:
I think "vehicles" is the wrong term.  It implies a containing relationship that these don't have. I'd just use "modes".